### ISSUE TITLE
Search attributes registration is lost after successfully adding

### ISSUE THREAD

**Issue by longquanzheng** [temporalio/temporal#4160]
## Expected Behavior
The search attributes should never got lost after registering 

## Actual Behavior
Sometimes it will get lost

## Steps to Reproduce the Problem
This is the script that I used to register search attributes automatically after service is up running in docker. 
It happens occasionally that one of the three (randomly) will be lost after the CLI command show up the registering is succeeded.

E.g. in this integ test failure: https://github.com/indeedeng/iwf/actions/runs/4682724673/jobs/8297177058

In dump docker logs you can see:
```
 You are about to add search attributes [IwfWorkflowType:Keyword]. Continue? Y/N yes
  Search attributes have been added
  You are about to add search attributes [IwfGlobalWorkflowVersion:Int]. Continue? Y/N yes
  Search attributes have been added
  You are about to add search attributes [IwfExecutingStateIds:Keyword]. Continue? Y/N yes
  Search attributes have been added
```

And then test fails:
```
ERROR	loggerimpl/logger.go:135	encounter error for API	{"error": "search attribute IwfWorkflowType is not defined", "logging-call-at": "service.go:389"}
```

## Specifications

  - Version: 1.20.1 , 1.19.1
  - Platform: docker

**Comment by rodrigozhou** [comment]
What persistence storage you're using? Can you run `temporal operator search-attribute list` to list the search attributes?

After adding search attributes, you may have to wait up to 1 min to be able to use it. The command above should list all search attributes though.

**Comment by rodrigozhou** [comment]
I checked your PR https://github.com/indeedeng/iwf/pull/245, and it shouldn't be necessary to add those `sleep`. Also, you can add multiple search attributes in a single call. Eg:
```
$ temporal search-attribute create --namespace default \
    --name CustomKeywordField --type Keyword \
    --name CustomIntField --type Int
```

**Comment by longquanzheng** [comment]
> I checked your PR https://github.com/indeedeng/iwf/pull/245, and it shouldn't be necessary to add those sleep

@rodrigozhou yeah the sleep doesn't always work...So I finally have to do a workaround using list and wait in https://github.com/indeedeng/iwf/pull/256
> you can add multiple search attributes in a single call
Good to know. Thanks! I will update it later.

**Comment by longquanzheng** [comment]
> What persistence storage you're using?

I am using docker see: 
https://github.com/indeedeng/iwf/blob/main/docker-compose/ci-temporal-dependencies.yml
https://github.com/indeedeng/iwf/blob/main/docker-compose/.env

**Comment by longquanzheng** [comment]
> After adding search attributes, you may have to wait up to 1 min to be able to use it

Interesting.. why it that? Is the list command enough to wait?

**Comment by rodrigozhou** [comment]
> Interesting.. why it that? Is the list command enough to wait?

There are caches that are refreshed once every minute. The list command bypasses the cache, and looks at the persistence directly, so it should have the up to date information.

**Comment by longquanzheng** [comment]
> > Interesting.. why it that? Is the list command enough to wait?
> 
> There are caches that are refreshed once every minute. The list command bypasses the caches, and looks at the persistence directly, so it should have the up to date information.

Actually, from what I observed it locally, the search attributes won’t show up after even hours , until I ran the command to re-register the same search attribute.

**Comment by rodrigozhou** [comment]
Do you have repro steps? Or logs from `tctl` and Server? I can try to investigate more.

**Comment by longquanzheng** [comment]
> Do you have repro steps? Or logs from `tctl` and Server? I can try to investigate more.

Sorry for late response
Thanks @jupudibhaskar967 for providing the repro steps. 
My repro steps would be just run the integ test in iWF without the workaround in https://github.com/indeedeng/iwf/pull/256 
One out of ten there should be a repro. But I think @jupudibhaskar967 's step would be much easier to follow

**Comment by jupudibhaskar967** [comment]
@rodrigozhou  Any updates on this?

**Comment by rodrigozhou** [comment]
@jupudibhaskar967 I found your issue. It happens when trying to add multiple search attributes of same type (eg: adding 2 keyword type search attributes). It might happen that one or more of them might be lost. This [PR](https://github.com/temporalio/temporal/pull/4273) should fix it.

@longquanzheng I recommend you to update your commands to use [Temporal CLI](https://github.com/temporalio/cli) instead of `tctl`. The latter will be deprecated eventually, and although we're trying to keep both in sync at this moment, we'll stop eventually. I just noticed that the code path of `tctl` was not up-to-date relative to Temporal CLI, and this might be the issue you're facing. I'm updating together with the PR above, and hopefully it will fix it.
